terratorch/models/smp_model_factory.py

```diff
@@ -1,83 +1,387 @@
-        return ModelOutput(smp_output)
+    This function constructs a decoder module based on the specified parameters and wraps it in a
+    custom wrapper that allows handling single or multiple embeddings. It also ensures that the
+    appropriate encoder parameters are passed and registered correctly.
 
-    def freeze_encoder(self):
-        raise NotImplementedError()
+    Args:
+        decoder (str): The name of the SMP decoder to use.
+        backbone_kwargs (dict): Dictionary of parameters for configuring the backbone.
+        smp_kwargs (dict): Dictionary of parameters specific to the SMP model.
+        aux_kwargs (dict): Dictionary of parameters for auxiliary configurations.
+        args (dict): Additional arguments, typically containing head-specific parameters.
+        out_channels (list[int] | int): The number of output channels for each layer of the decoder.
+        in_channels (int): The number of input channels.
+        num_classes (int): The number of output classes for the model.
+        output_stride (int): The output stride of the decoder.
 
-    def freeze_decoder(self):
-        raise NotImplementedError()
+    Returns:
+        SMPDecoderForPrithviWrapper: A wrapped decoder module configured based on the provided parameters.
+
+    Raises:
+        ValueError: If the specified decoder is not supported by SMP.
+    """
+
+    # Gets decoder class.
+    decoder = decoder.removeprefix("smp_")
+    decoder_module = getattr(smp, decoder, None)
+    if decoder_module is None:
+        msg = f"Decoder {decoder} is not supported in SMP."
+        raise ValueError(msg)
+
+    # head args for handling embeds.
+    head_kwargs = _extract_prefix_keys(args, "head_")
+
+    # Little hack to make SMP model accept our encoder.
+    # passes a dummy encoder to be changed later.
+    # this is needed to pass encoder params.
+    backbone_kwargs["out_channels"] = out_channels
+    backbone_kwargs["output_stride"] = output_stride
+    aux_kwargs = None if aux_kwargs == {} else aux_kwargs
+
+    dummy_encoder = _make_smp_encoder()
+
+    _register_custom_encoder(dummy_encoder, backbone_kwargs, None)
+
+    dummy_encoder = dummy_encoder(
+        depth=smp_kwargs["encoder_depth"],
+        output_stride=backbone_kwargs["output_stride"],
+        out_channels=backbone_kwargs["out_channels"],
+    )
+
+    model_args = {
+        "encoder_name": "SMPEncoderWrapperWithPFFIM",
+        "encoder_weights": None,
+        "in_channels": in_channels,
+        "classes": num_classes,
+        **smp_kwargs,
+    }
+
+    # Creates model with dummy encoder and decoder.
+    model = decoder_module(**model_args, aux_params=aux_kwargs)
+
+    # Wrapper for SMP Decoder.
+    smp_decoder = SMPDecoderForPrithviWrapper(decoder=model.decoder, num_channels=out_channels[-1])
+    if "multiple_embed" in head_kwargs:
+        smp_decoder.forward = smp_decoder.forward_multiple_embeds
+    else:
+        smp_decoder.forward = smp_decoder.forward_single_embed
```

Comment:
see suggestion above to allow `in_index` parameter, which would make this monkey patching not necessary